libs/vm/src/analyser.cpp

```diff
@@ -357,25 +390,438 @@ bool Analyser::Analyse(BlockNodePtr const &root, std::vector<std::string> &error
+    }
+    if (HasAnnotation(annotations, QUERY_ANNOTATION))
+    {
+      is_query = true;
+    }
+    if (HasAnnotation(annotations, CLEAR_ANNOTATION))
+    {
+      metadata.clear_functions.emplace_back(function_node, filename_);
+    }
+    if (HasAnnotation(annotations, OBJECTIVE_ANNOTATION))
+    {
+      metadata.objective_functions.emplace_back(function_node, filename_);
+    }
+    if (HasAnnotation(annotations, PROBLEM_ANNOTATION))
+    {
+      metadata.problem_functions.emplace_back(function_node, filename_);
+    }
+    if (HasAnnotation(annotations, WORK_ANNOTATION))
+    {
+      metadata.work_functions.emplace_back(function_node, filename_);
+    }
+  }
+  break;
+  case NodeKind::MemberFunctionDefinition:
+  {
+    if (annotations)
+    {
+      AddError(annotations->line, "Annotations on member functions are not permitted");
+    }
+  }
+  break;
+  default:
+    assert(false);
+    break;
+  }
+
+  auto const &return_type = function_name_node->function->return_type != nullptr
+                                ? function_name_node->function->return_type
+                                : void_type_;
+
+  if (is_action)
+  {
+    if (return_type != int64_type_ && return_type != void_type_)
+    {
+      AddError(function_name_node->line, "@action functions must either be void or return Int64");
+    }
+  }
+
+  if (is_init)
+  {
+    if (return_type != int64_type_ && return_type != void_type_)
+    {
+      AddError(function_name_node->line, "@init functions must either be void or return Int64");
+    }
+
+    auto const param_types = function_name_node->function->parameter_types;
+    if (!param_types.empty() && !(param_types.size() == 1 && param_types.back() == address_type_))
+    {
+      AddError(function_name_node->line,
+               "@init functions must accept eithoer no parameters or exactly one parameter of type "
+               "Address");
+    }
+  }
+
+  if (is_query)
+  {
+    if (return_type == void_type_)
+    {
+      AddError(function_name_node->line, "@query functions must not be void");
+    }
+  }
+}
+
+void Analyser::CheckInitFunctionUnique(LedgerRestrictionMetadata const &metadata)
+{
+  if (metadata.init_functions.size() > 1)
+  {
+    for (auto const &x : metadata.init_functions)
+    {
+      AddError(x.filename, x.node->line,
+               "Multiple @init functions found; only one @init function is allowed per contract");
+    }
+  }
+}
+
+bool Analyser::CheckSynergeticFunctionsPresentAndUnique(LedgerRestrictionMetadata const &metadata)
+{
+  if (metadata.clear_functions.empty() && metadata.objective_functions.empty() &&
+      metadata.problem_functions.empty() && metadata.work_functions.empty())
+  {
+    return false;
+  }
+
+  if (metadata.clear_functions.size() == 1 && metadata.objective_functions.size() == 1 &&
+      metadata.problem_functions.size() == 1 && metadata.work_functions.size() == 1)
+  {
+    return true;
+  }
+
+  bool const clear_missing     = metadata.clear_functions.empty();
```

Comment:
Pedantic, but maybe put these bools at the top and reuse?

Reply:
done